Add a REST layer on top of the feature store

nomoa opened the pull request:
Added the following:
- Simple create/delete operations for manipulating stores.
- Simple CRUD operations for feature, featureset and model.
- Clear cache operation
- Fetch cluster wide cache stats
- Create/update a featureset from existing features
- Create a model from an existing featureset

Stores are not created automatically and must be created before using them.
Features and featureset can be updated, models can only be created.

closes #34
closes #33

nomoa commented:
@ebernhardson, @gehel a new giant patch to review, sorry! :(

gehel commented on src/main/java/com/o19s/es/ltr/action/TransportAddFeatureToSetAction.java:
```diff
@@ -0,0 +1,243 @@
+        private void onActionFailure(Exception e) {
+            exception.set(e);
+            if (countdown.countDown()) {
+                listener.onFailure(exception.get());
```
Here, exception.get() != e (at least some of the time if both search and get are in error). This is probably a mistake, but I'm not entirely sure of the context.

nomoa replied:
refactored

gehel commented on src/main/java/com/o19s/es/ltr/action/TransportAddFeatureToSetAction.java:
```diff
@@ -0,0 +1,243 @@
+
+        private void mergeAndStore() {
+            if (exception.get() != null) {
+                listener.onFailure(exception.get());
```
Don't we notify the listener twice of that exception? Once in onActionFailure(), once here? Also, what happens if there is more than one exception?

nomoa replied:
refactored

gehel commented on src/main/java/com/o19s/es/ltr/action/TransportClearCachesAction.java:
```diff
@@ -0,0 +1,124 @@
+    protected ClearCachesNodeResponse nodeOperation(ClearCachesNodeRequest request) {
+        ClearCachesNodesRequest r = request.request;
+        switch (r.getOperation()) {
+        case ClearStore: {
```
Code blocks in switch statements? Syntactically correct, but I know only one person who uses code blocks... (well, now two).

nomoa replied:
done

gehel commented on src/main/java/com/o19s/es/ltr/action/TransportFeatureStoreAction.java:
```diff
@@ -0,0 +1,138 @@
+
+    @Override
+    protected void doExecute(FeatureStoreRequest request, ActionListener<FeatureStoreResponse> listener) {
+        logger.warn("attempt to execute a TransportFeatureStoreAction without a task");
```
logging AND throwing exception is usually redundant. I would expect the caller to log exception as well, so we will probably end up with duplicate log messages (well, the messages will be slightly different, but the point is that we could have all the context in the exception and remove the logger.warn() call). That same pattern is used in a few other places.

nomoa replied:
done

gehel commented on docs/rest_api.md:
```diff
@@ -0,0 +1,147 @@
+### Append features to a set
+A set can also be created/updated from existing features:
+
+`POST /_ltr/_featureset/my_featureset/_addfeatures/my_feature*`
```
After discussion with @nomoa, this operation actually fails the second time as the features have already been added. POST make sense.

:+1:

nomoa commented on src/main/java/com/o19s/es/ltr/LtrQueryParserPlugin.java:
```diff
@@ -17,40 +17,85 @@
+    }
+
+    public static LtrRankerParserFactory buildParserFactory() {
+        Supplier<RankerFactory> ranklib = Suppliers.memoize(RankerFactory::new);
```
added a comment

nomoa commented on src/main/java/com/o19s/es/ltr/action/TransportAddFeatureToSetAction.java:
```diff
@@ -0,0 +1,243 @@
+            srequest.source().query(bq);
+            srequest.source().fetchSource(true);
+            srequest.source().size(StoredFeatureSet.MAX_FEATURES);
+            searchAction.execute(srequest, wrap(this::onSearchResponse, this::onActionFailure));
```
done

nomoa commented on src/main/java/com/o19s/es/ltr/action/CachesStatsAction.java:
```diff
@@ -0,0 +1,246 @@
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.field("all", all);
+            builder.startArray("nodes");
```
makes sense, fixing

ebernhardson commented:
I was able to load in some features, create a feature set, add an xgboost model (using followup patch), and run an "sltr" query against it. According to the explain this generated appropriate feature vectors, although for some reason all my results get the same score, but this might simply be because the model was trained against a full enwiki index and i have a tiny index of 200 documents locally, so the split values might all be useless.

ebernhardson commented:
I was able to load in a dozen features, build a feature set out of them, and then with the xgboost json followup patch load in an xgboost model. I was then able to execute an "sltr" query and get results, although ever result has the same score. This is very likely because the index trained against and the index i'm testing against are completely different, so probably nothing to worry about yet.

ebernhardson commented:
Above problem with same values for everything was indeed just an unrepresentative index. I pulled in 100k docs from one of the dumps and i now get varied scores for documents.

ebernhardson commented:
It might also be nice to have some basic information on a few endpoints that are currently unclaimed:

GET /_ltr - Perhaps a list of stores available? not sure
GET /_ltr/_feature - List of features?
GET /_ltr/_featureset - List of named feature sets?
GET /_ltr/_model - List of models?

nomoa commented on src/main/java/com/o19s/es/ltr/action/TransportCreateModelFromSetAction.java:
```diff
@@ -0,0 +1,102 @@
+    private void doStore(Task parentTask, GetResponse response, CreateModelFromSetRequest request,
+                         ActionListener<CreateModelFromSetResponse> listener) {
+        if (!response.isExists()) {
+            throw new IllegalArgumentException("Stored feature [" + request.getFeatureSetName() + "] does not exist");
```
done

nomoa commented on src/main/java/com/o19s/es/ltr/action/TransportCreateModelFromSetAction.java:
```diff
@@ -0,0 +1,102 @@
+            throw new IllegalArgumentException("Stored feature [" + request.getFeatureSetName() + "] does not exist");
+        }
+        if (request.getExpectedSetVersion() != null && request.getExpectedSetVersion() != response.getVersion()) {
+            throw new IllegalArgumentException("Stored feature [" + request.getFeatureSetName() + "]" +
```
done

nomoa commented on docs/rest_api.md:
```diff
@@ -0,0 +1,147 @@
+
+### Models
+
+`/_ltr/_model/feature_name` supports `PUT`, `GET` and `DELETE` operations.
```
done

nomoa commented on docs/rest_api.md:
````diff
@@ -0,0 +1,147 @@
+Get cluster wide cache statistics:
+```
+GET /_ltr/_cachestats
+```
````
Added a small note under the Feature Store section

nomoa commented on src/main/java/com/o19s/es/ltr/action/FeatureStoreAction.java:
```diff
@@ -0,0 +1,216 @@
+                arve = addValidationError("storable element must be set", arve);
+            }
+            if (action == Action.UPDATE && !storableElement.updatable()) {
+                arve = addValidationError("Element of type [" + storableElement.type() + "] are not updatable.", arve);
```
done

nomoa commented on src/main/java/com/o19s/es/ltr/rest/RestCreateModelFromSet.java:
```diff
@@ -0,0 +1,101 @@
+        PARSER = new ObjectParser<>("model", ParserState::new);
+        PARSER.declareString(ParserState::setName, new ParseField("name"));
+        PARSER.declareString(ParserState::setModelType, new ParseField("type"));
+        PARSER.declareString(ParserState::setModelDefinition, new ParseField("definition"));
```
makes sense, not sure how to do that cleanly yet.
I'll try to change the REST api for this and use the Content-Type http header to read the model name.
Doing this while parsing is bit tricky since I'd need to change the model representation from String to either String/XContent ...

nomoa commented on src/main/java/com/o19s/es/ltr/action/ClearCachesAction.java:
```diff
@@ -0,0 +1,177 @@
+                arve = addValidationError("no operation provided", arve);
+            }
+
+            if (operation != null && operation != Operation.ClearStore && name == null) {
```
in fact it can, we do not fail on the previous if and here we continue to collect validation errors so we must check against null not to add a second error.

nomoa commented:
@ebernhardson I think I addressed all your comments, in the end I did like you suggested concerning json models: I do exactly the same as I do for feature with a `modelAsString` flag.
I added a simple LinearRankerParser for easy testing.

@gehel if you have 5 mins I'd love if you could check the last 3 patches, thanks! :)

gehel commented on src/main/java/com/o19s/es/ltr/rest/RestSimpleFeatureStore.java:
```diff
@@ -0,0 +1,249 @@
+        int size = request.paramAsInt("size", 20);
+        BoolQueryBuilder qb = QueryBuilders.boolQuery().filter(QueryBuilders.termQuery("type", type));
+        if (prefix != null && !prefix.isEmpty()) {
+            qb.must(QueryBuilders.matchQuery("name.prefix", prefix));
```
A static import of QueryBuilders.matchQuery would make this read more fluently (very minor, feel free to ignore).

gehel commented on src/main/java/com/o19s/es/ltr/feature/store/index/IndexFeatureStore.java:
```diff
@@ -53,44 +92,170 @@ public String getStoreName() {
+     * Construct the elasticsearch index name based on a store name
+     */
+    public static String indexName(String storeName) {
+        if (Objects.requireNonNull(storeName).isEmpty()) {
```
We throw a NPE with no message if `storeName` is `null` but an `IllegalArgumentException` if it is empty. This sounds like a job for Guava's `Preconditions.checkArgument()` and `Strings.isNullOrEmpty()`, but we don't have it as a dependency yet (probably not worth adding it, so feel free to ignore).

nomoa replied:
I'll ignore that one :)

gehel commented on src/test/java/com/o19s/es/ltr/feature/store/index/IndexFeatureStoreTests.java:
```diff
@@ -0,0 +1,124 @@
+        assertTrue(isIndexStore(IndexFeatureStore.DEFAULT_STORE));
+        assertFalse(isIndexStore("not_really"));
+        assertFalse(isIndexStore(IndexFeatureStore.STORE_PREFIX));
+        for (int i = 0; i < random().nextInt(10) + 10; i++) {
```
`random().nextInt(10) + 10` is evaluated for each step of the loop. It is technically correct, but it feels a bit strange to have a non constant bound to a loop (feel free to ignore).

Same comment applies to `testStoreName()`.

nomoa replied:
doh I'm stupid, fixing...

ebernhardson commented:
I don't know if it needs to go here, this patch is big enough, but i've been trying to debug getting a no match for all templated features and it might be useful to have some sort of debug endpoint that can report back what the query generated for a feature is.

nomoa commented on src/main/java/com/o19s/es/ltr/ranker/parser/XGBoostJsonParser.java:
```diff
@@ -0,0 +1,168 @@
+
+        Node toNode(FeatureSet set) {
+            if (isSplit()) {
+                return new NaiveAdditiveDecisionTree.Split(children.get(1).toNode(set), children.get(0).toNode(set),
```
sorry!!
reading back your comments on my fork I completely missed your first line about this issue...
Fixing...
I think I'm confused by the meaning of a successful condition:
for me a condition is met if `feature_score > split.threshold` => `right` but it's `feature_score < split.threshold` => `left`.
So for me `yes` was `go right`...
We don't really care but `missing` => `no` while it would make more sense to go to `yes`?
Well... I'll ignore completely `missing`... I have added some code to constraint the structure but since we don't use it it's confusing to enforce a particular side for `missing`.

gehel commented:
From my limited understanding, those last 2 commits look good to me...

nomoa commented:
concerning missing params I default to `match_none` if one of them is missing.
That is probably misleading, I could let the user specify a default query to use if missing and fail if no default is set?

nomoa commented:
NOTE: a version of the plugin using this branch is being deployed to our production clusters for load testing and possibly a/b test.
merging so that it'll be easier to track future changes.
